### PR TITLE
DAOS-13471 control: Add JSON-format version to utilities

### PR DESCRIPTION
**Pull request by mjmac**
Allow consumers to grab structured build/version information.
Centralizes JSON output logic to a single implementation in
the cmdutil package. Includes updates to control/version.py
to use the structured version information instead of scraping
stdout.

Signed-off-by: Michael MacDonald <mjmac.macdonald@intel.com>

### CODE REVIEW

**Review by daltonbohning — COMMENTED** [review]
ftest LGTM.

File: src/control/cmd/daos_server/main.go [hunk 1]
```diff
@@ -144,11 +164,11 @@ func parseOpts(args []string, opts *mainOpts, log *logging.LeveledLogger) error
 
 	// Parse commandline flags which override options loaded from config.
 	_, err := p.ParseArgs(args)
-	if err != nil {
-		return err
+	if opts.JSON && wroteJSON.IsFalse() {
+		cmdutil.OutputJSON(os.Stdout, nil, err)
```
Comment (kjacque) [inline src/control/cmd/daos_server/main.go:168]: same comment as above

**Review by kjacque — APPROVED** [review]
Overall LGTM. Two minor comments.

File: src/control/cmd/daos_agent/main.go [hunk 2]
```diff
@@ -257,11 +239,12 @@ func parseOpts(args []string, opts *cliOptions, invoker control.Invoker, log *lo
 			ctlCmd.setInvoker(invoker)
 		}
 
-		if err := cmd.Execute(args); err != nil {
-			return err
+		err = cmd.Execute(args)
+		if opts.JSON && wroteJSON.IsFalse() {
+			cmdutil.OutputJSON(os.Stdout, nil, err)
```
[the remark this answers is not in the document]
Comment (mjmac) [reply src/control/cmd/daos_agent/main.go:244]: This is the only opportunity to emit JSON output if there was an error. In `main()`, we don't have access to the `wroteJSON` bool to determine whether or not JSON was already emitted somewhere else.

File: src/tests/ftest/control/version.py [hunk 3]
```diff
@@ -41,66 +50,23 @@ def test_version(self):
...
-        if "DAOS Agent" not in stdout:
-            errors.append("DAOS Agent is not in the output! {}".format(stdout))
-
-        result = re.findall(r"DAOS Agent v([\d.]+)", stdout)
-        if not result:
-            errors.append("Failed to obtain daos_agent version! {}".format(output))
-        else:
-            daos_agent_version = result[0]
-            self.log.info("daos_agent version = %s", daos_agent_version)
+        daos_agent_version = json.loads("".join(output[0]["stdout"]))["response"]["version"]
```
Comment (knard-intel) [inline src/tests/ftest/control/version.py:63]: The field `exit_status` should probably be checked before accessing the `output` one.

**Review by knard-intel — APPROVED** [review]
Minor remarks,
I was also working on the `src/tests/ftest/control/version.py` file with my ongoing PR  #12570.
Your solution should be more robust as it is based on json formating.
If it is OK for you, I will wait after your merge to integrate some of my modifications of this file.

**Pull request merged**